sync2/p2p.go

```diff
@@ -0,0 +1,168 @@
+func DefaultConfig() Config {
+	return Config{
+		RangeSetReconcilerConfig:  rangesync.DefaultConfig(),
+		MultiPeerReconcilerConfig: multipeer.DefaultConfig(),
+		TrafficLimit:              200_000_000,
+		MessageLimit:              20_000_000,
+	}
+}
+
+// P2PHashSync is handles the synchronization of a local OrderedSet against other peers.
+type P2PHashSync struct {
+	logger     *zap.Logger
+	cfg        Config
+	os         rangesync.OrderedSet
+	syncBase   multipeer.SyncBase
+	reconciler *multipeer.MultiPeerReconciler
+	cancel     context.CancelFunc
+	eg         errgroup.Group
+	start      sync.Once
+	running    atomic.Bool
+}
+
+// NewP2PHashSync creates a new P2PHashSync.
+func NewP2PHashSync(
+	logger *zap.Logger,
+	d *Dispatcher,
+	name string,
+	os rangesync.OrderedSet,
+	keyLen, maxDepth int,
+	peers *peers.Peers,
+	handler multipeer.SyncKeyHandler,
+	cfg Config,
+	requester rangesync.Requester,
+) *P2PHashSync {
+	s := &P2PHashSync{
+		logger: logger,
+		os:     os,
+		cfg:    cfg,
+	}
+	// var ps multipeer.PairwiseSyncer
+	ps := rangesync.NewPairwiseSetSyncer(logger, requester, name, cfg.RangeSetReconcilerConfig)
+	s.syncBase = multipeer.NewSetSyncBase(logger, ps, s.os, handler)
+	s.reconciler = multipeer.NewMultiPeerReconciler(
+		logger, cfg.MultiPeerReconcilerConfig,
+		s.syncBase, peers, keyLen, maxDepth)
+	d.Register(name, s.serve)
+	return s
+}
+
+func (s *P2PHashSync) serve(ctx context.Context, stream io.ReadWriter) error {
+	peer, found := server.ContextPeerID(ctx)
+	if !found {
+		panic("BUG: no peer ID found in the handler")
+	}
+	// We derive a dedicated Syncer for the peer being served to pass all the received
+	// items through the handler before adding them to the main ItemStore
+	return s.syncBase.Derive(peer).Serve(ctx, stream)
+}
+
+// Set returns the OrderedSet that is being synchronized.
+func (s *P2PHashSync) Set() rangesync.OrderedSet {
+	return s.os
+}
+
+// Load loads the OrderedSet from the underlying storage.
+func (s *P2PHashSync) Load() error {
+	s.logger.Info("loading the set")
+	start := time.Now()
+	// We pre-load the set to avoid waiting for it to load during a
+	// sync request
+	if err := s.os.EnsureLoaded(); err != nil {
+		return fmt.Errorf("load set: %w", err)
+	}
+	info, err := s.os.GetRangeInfo(nil, nil)
+	if err != nil {
+		return fmt.Errorf("get range info: %w", err)
+	}
+	s.logger.Info("done loading the set",
+		zap.Duration("elapsed", time.Since(start)),
+		zap.Int("count", info.Count),
+		zap.Stringer("fingerprint", info.Fingerprint))
+	return nil
+}
+
+// Start starts the multi-peer reconciler.
+func (s *P2PHashSync) Start() {
+	if !s.cfg.EnableActiveSync {
+		s.logger.Info("active sync is disabled")
+		return
+	}
+	s.running.Store(true)
+	s.start.Do(func() {
+		s.eg.Go(func() error {
+			defer s.running.Store(false)
+			var ctx context.Context
+			ctx, s.cancel = context.WithCancel(context.Background())
+			return s.reconciler.Run(ctx)
+		})
```

Comment:
`Start` just serves as a wrapper around the `Run` method here. Would it make sense to instead of having the `Start`,  `Stop` and `s.running` methods/fields to just have a `Run` method that passes along the context to `s.reconciler`? This would also get rid of the need for `s.cancel`.

Reply:
`P2PHashSync` has active sync enable/disable flag as its config option. That's part of `P2PHashSync` logic.
When `cfg.EnableActiveSync` is `false`, `Start` / `Stop` are noops. In this case, `P2PHashSync` only serves requests received from a p2p `Server` via a `Dispatcher`.